oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java

```diff
@@ -81,47 +85,65 @@ public void setAllowFormEncodedBodyParameter(boolean allowFormEncodedBodyParamet
 	 *
 	 * @param allowUriQueryParameter if the URI query parameter is supported
 	 */
-	public void setAllowUriQueryParameter(boolean allowUriQueryParameter) {
+	public void setAllowUriQueryParameter(final boolean allowUriQueryParameter) {
 		this.allowUriQueryParameter = allowUriQueryParameter;
 	}
 
-	private static String resolveFromAuthorizationHeader(HttpServletRequest request) {
-		String authorization = request.getHeader(HttpHeaders.AUTHORIZATION);
-		if (StringUtils.hasText(authorization) && authorization.startsWith("Bearer")) {
-			Matcher matcher = authorizationPattern.matcher(authorization);
+	private void validate(final HttpServletRequest request, final String authorizationHeaderToken) {
+		if (hasMultipleAccessTokens(request, authorizationHeaderToken)) {
+			throw new OAuth2AuthenticationException(BEARER_TOKEN_ERROR_MULTIPLE);
+		}
+		if (this.allowFormEncodedBodyParameter == true) {
+			if (this.allowUriQueryParameter == false && isTokenInUriQueryParameter(request)) {
+				throw new OAuth2AuthenticationException(BEARER_TOKEN_ERROR_MALFORMED);
+			}
+		}
+	}
 
+	private String resolveFromAuthorizationHeader(final HttpServletRequest request) {
```

Comment:
Adding `final` allows compiler to do certain optimizations as well also reduces bugs due to accidental changes to value.

Reply:
The only real use for `final` parameters was when dealing with anonymous inner classes and that was removed in Java 8 with introduction of effectively final.

Reply:
@vpavic  I disagree on real usage you mentioned, as per me real usage is programmer's intent that the value must not be changed (immutability).

I do agree it adds more changes in PR hence additional work to review as well as it may have implications on back patch hence I will revert this change.